stub/src/main/java/io/grpc/stub/ClientCalls.java

```diff
@@ -123,25 +123,30 @@ private ClientCalls() {}
   public static <ReqT, RespT> RespT blockingUnaryCall(
       Channel channel, MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, ReqT req) {
     ThreadlessExecutor executor = new ThreadlessExecutor();
+    boolean interrupt = false;
     ClientCall<ReqT, RespT> call = channel.newCall(method, callOptions.withExecutor(executor));
     try {
       ListenableFuture<RespT> responseFuture = futureUnaryCall(call, req);
       while (!responseFuture.isDone()) {
         try {
           executor.waitAndDrain();
         } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw Status.CANCELLED
-              .withDescription("Call was interrupted")
-              .withCause(e)
-              .asRuntimeException();
+          interrupt = true;
```

Comment:
optional:

```
...
int interrupt = 0;
if (interrupts++ == 1) {
  throw Status.....asRuntimeException();
}
```

This allows a single interrupt to clean up, but a second interrupt while waiting for onClose to bypass.

Reply:
Cancelling multiple times is a noop. I'm not too interested in optimizing this, especially since multiple interrupts would be unlikely.

Reply:
It's not so much an optimization, but a safety check.  If the onClose callback gets dropped it would allow this thread to escape and shutdown.

Reply:
We allow an infinite number of interrupts, waiting for the onClose. I'm not sure how the onClose could be dropped here.

Reply:
We discussed offline, and the thought is this could block for a long time waiting for the onClose. If interrupted twice we could choose to exit early, without waiting for the onClose.

I think this is more appropriate at the moment, although I agree it is unfortunate we have to wait for the onClose. If we had #636 done and a fallback executor, we could add a shutdown to ThreadlessExecutor where it would begin throwing RejectedExecutionException on execute(). We would then shutdown() when we exit the loop here. We can then return more quickly, knowing that if/when the onClose is scheduled it will still be run (on the fallback executor).